src/EllCrv/Finite.jl

```diff
@@ -816,3 +817,161 @@ function trace_of_frobenius(E::EllCrv{T}, n::Int) where T<:FinFieldElem
   end
 end
 
+################################################################################
+#
+#  Supersingular Elliptic Curves
+#
+################################################################################
+
+
+#Following Identifying supersingular elliptic curves - Andrew V. Sutherland
+@doc Markdown.doc"""
+    is_supersingular(E::EllCrv{T}) where T <: FinFieldElem
+Return true when the elliptic curve is supersingular. The result is proven to be correct.
+"""
+function is_supersingular(E::EllCrv{T}) where T <: FinFieldElem
+  K = base_field(E)
+  
+  p = characteristic(K)
+  j = j_invariant(E)
+  
+  if j^(p^2) != j
+    return false
+  end
+  
+  if p<= 3
+    return j == 0
+  end
+  
+  L = GF(p, 2)
+  Lx, X = PolynomialRing(L, "X")
+  Lxy, Y = PolynomialRing(Lx, "Y")
+  Phi2 = X^3 + Y^3 - X^2*Y^2 + 1488*(X^2*Y + Y^2*X) - 162000*(X^2 + Y^2) + 40773375*X*Y + 8748000000*(X + Y) - 157464000000000
+  
+  jL = _embed_into_p2(j, L)
+  
+  js = roots(Phi2(jL))
+  
+  if length(js) < 3
+    return false
+  end
+  
+  newjs = [jL, jL, jL]
+  f = elem_type(Lx)[zero(Lx), zero(Lx), zero(Lx)]
+  
+  m = nbits(p) - 1
+  for k in (1 : m)
+    for i in (1 : 3)
+      f[i] = divexact(Phi2(js[i]), X - newjs[i])
+      newjs[i] = js[i]
+      froots = roots(f[i])
+      if isempty(froots)
+        return false
+      end
+      js[i] = froots[1]
+    end
+  end
+  return true
+end
+
+function _to_z(a::Union{gfp_elem, gfp_fmpz_elem})
+  return lift(a)
+end
+
+function _to_z(a::Union{fq_nmod, fq})
+  return coeff(a, 0)
+end
+
+function _embed_into_p2(j, L)
+  K = parent(j)
+  # The easy case
+  if degree(K) == 1
+    return L(_to_z(j))
+  else
+    p = minpoly(j)
+    # Easy case
+    if degree(p) <= 1
+      return L(_to_z(j))
+    end
+    F, a = FiniteField(p)
+    e = embed(F, L)
+    return e(gen(F))
+  end
+end
+
+@doc Markdown.doc"""
+    is_ordinary(E::EllCrv{T}) where T <: FinFieldElem
+Return true when the elliptic curve is ordinary, i.e. not supersingular.
+"""
+function is_ordinary(E::EllCrv{T}) where T <: FinFieldElem
+  return !is_supersingular(E)
+end
+
+#Following Identifying supersingular elliptic curves - Andrew V. Sutherland
+@doc Markdown.doc"""
+    is_probable_supersingular(E::EllCrv{T}) where T <: FinFieldElem
+Uses a probabilistic algorithm to test whether E is supersingular or not.
+If the function returns false, the curve is proven to be ordinary.
+If the function returns true, there is a high chance the curve is supersingular,
+but the result hasn't been proven.
+"""
+function is_probable_supersingular(E::EllCrv{T}) where T <: FinFieldElem
+  j = j_invariant(E)
+  K = base_field(E)
+  p = characteristic(K)
+  
+  local degj::Int
+
+  if degree(K) == 1
+    degj = 1
+  else
+    degj = degree(minpoly(j))
+  end
+  
+  if degj == 1
+    return monte_carlo_test(E, p+1)
+  elseif degj == 2
+    return monte_carlo_test(E, p+1) || monte_carlo_test(E, p+1)
```

Comment:
Stupid question, but should the second be `p-1`?

Reply:
It should be. I changed it.